tests/test_package_swap.py

```diff
@@ -0,0 +1,262 @@
+                                      products_info=self.pack_4apples_info,
+                                      confirm=True,
+                                      assign=True)
+
+        with self.assertRaises(ValidationError):
+            picking.update_picking(package_name=SCANNED_PACKAGE_NAME,
+                                   expected_package_name=EXPECTED_PACKAGE_NAME)
+
+    def test06_error_if_different_package_product(self):
+        """ Should error if package products are different """
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.expected_package.id)
+        self.create_quant(self.banana.id, self.test_location_01.id, 4,
+                          package_id=self.scanned_package.id)
+        picking = self.create_picking(self.picking_type_pick,
+                                      products_info=self.pack_4apples_info,
+                                      confirm=True,
+                                      assign=True)
+
+        with self.assertRaises(ValidationError):
+            picking.update_picking(package_name=SCANNED_PACKAGE_NAME,
+                                   expected_package_name=EXPECTED_PACKAGE_NAME)
+
+    def test07_error_if_different_package_qty(self):
+        """ Should error if products' qty are different """
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.expected_package.id)
+        self.create_quant(self.apple.id, self.test_location_01.id, 8,
+                          package_id=self.scanned_package.id)
+        picking = self.create_picking(self.picking_type_pick,
+                                      products_info=self.pack_4apples_info,
+                                      confirm=True,
+                                      assign=True)
+
+        with self.assertRaises(ValidationError):
+            picking.update_picking(package_name=SCANNED_PACKAGE_NAME,
+                                   expected_package_name=EXPECTED_PACKAGE_NAME)
+
+    def test08_error_if_different_picking_type(self):
+        """
+        Should error when swapping two packages:
+         - with same contents;
+         - in the same location;
+         - both scanned and expected packages are reserved, but
+           with different picking types.
+
+        """
+        PickingType = self.env['stock.picking.type']
+
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.expected_package.id)
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.scanned_package.id)
+        picking = self.create_picking(self.picking_type_pick,
+                                      products_info=self.pack_4apples_info,
+                                      confirm=True,
+                                      assign=True)
+
+        other_picking_type = \
```

Comment:
get it from the warehouse :) check if it is in self already, we have used it in other tests.